source/common/upstream/load_balancer_impl.h

```diff
@@ -23,15 +23,32 @@ static constexpr uint32_t kOverProvisioningFactor = 140;
 /**
  * Base class for all LB implementations.
  */
-class LoadBalancerBase {
+class LoadBalancerBase : public LoadBalancer {
 public:
   // A utility function to chose a priority level based on a precomputed hash and
   // a priority vector in the style of per_priority_load_
   //
   // Returns the priority, a number between 0 and per_priority_load.size()-1
   static uint32_t choosePriority(uint64_t hash, const std::vector<uint32_t>& per_priority_load);
 
+  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+
 protected:
+  /**
+   * By implementing this method instead of chooseHost, host selection will
+   * be be subject to host filters specified by LoadBalancerContext.
+   *
+   * Host selection will be retried up to the number specified by
+   * hostSelectionRetryCount on LoadBalancerContext, and if no hosts are found
+   * within the allowed attempts, the host that was selected during the last
+   * attempt will be returned.
+   *
+   * If host selection is not random (i.e. retrying will not change the outcome),
```

Comment:
I think that "not random" is a bit too strong here; I'd phrase it as "If host selection is idempotent".